dg.go

```diff
@@ -265,14 +265,22 @@ func (n *node[NodeType]) Item() NodeType {
 }
 
 func (n *node[NodeType]) ResolutionStatus() ResolutionStatus {
-	// The status can change and be accessed on different threads, so lock to prevent races.
+	// Since the status can be changed by another thread, we access it under a
+	// lock here to prevent Go from reporting a data race; however, the actual
+	// status may change as soon as the lock is released, and so the return
+	// value, unless it is a terminal value, does not necessarily represent the
+	// current value.
```

Comment:
Again, it's much more than just "the actual status may change". Your lock acts here as just a memory barrier. That means that if the other thread does "morph node ; lock or memory barrier ; set ready or resolution status" this code is guaranteed to see the "morph node" results when it sees the new status, but if the other thread does "lock ; set ready and morph node ; unlock" and the "observer" uses `if node.IsReady() { observe node }`, the "observe node" code is *not* guaranteed to see the "morph node" results. (Worse, it's not guaranteed *not to see it*, because there *is* a memory barrier in the unlock, but you're leaving a small window where the change to `ready` may randomly become visible before the other memory writes.)

That is, it needs to either be "lock ; set node state including ready ; unlock" and "lock ; if ready { observe node } ; unlock" *or* "set node state ; write-barrier ; set ready" and "if ready { read-barrier ; observe node }". (And I wouldn't recommend the latter unless concurrent performance is a critical consideration, which I don't believe it is -- the lock is a lot easier and safer because it's less likely to be corrupted by some later developer who's not fully aware of the requirements.)